File: frontend/src/lib/schemas/ticker.ts
import { z } from 'zod';

export const tickerSearchResultSchema = z.object({
  ticker: z.string(),
  name: z.string(),
  market: z.string(),
  type: z.string(),
});

export const tickerSchema = z.object({
  ticker: z.string().min(1).max(20),
  company_name: z.string().min(1),
  shares: z.coerce.number().gt(0),
  /** Optional cluster assignment at creation time. */
  cluster_id: z.number().nullable().optional(),
});

export const tickerUpdateSchema = z.object({
  shares: z.coerce.number().gt(0),
  cluster_id: z.number().nullable().optional(),
});

export const tickerResponseSchema = z.object({
  id: z.number(),
  ticker: z.string(),
  company_name: z.string(),
  shares: z.coerce.number(),
  // Cluster assignment — null/undefined if unassigned
  cluster_id: z.number().nullable().optional(),
  // Market-data fields — null/undefined until first sync
  current_price: z.coerce.number().nullable().optional(),
  previous_close: z.coerce.number().nullable().optional(),
  day_change: z.coerce.number().nullable().optional(),
  day_change_pct: z.coerce.number().nullable().optional(),
  position_value: z.coerce.number().nullable().optional(),
  beta: z.coerce.number().nullable().optional(),
  synced_at: z.string().nullable().optional(),
  created_at: z.string(),
  updated_at: z.string(),
});

export type TickerSearchResult = z.infer<typeof tickerSearchResultSchema>;
export type Ticker = z.infer<typeof tickerSchema>;
export type TickerUpdate = z.infer<typeof tickerUpdateSchema>;
export type TickerResponse = z.infer<typeof tickerResponseSchema>;

File: backend/src/atlas/schemas/ticker.py
"""Pydantic schemas for portfolio tickers and Polygon ticker search."""

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, field_validator


class TickerSearchResult(BaseModel):
    """One result from the Polygon.io ticker-search endpoint."""

    ticker: str
    name: str
    market: str
    type: str


class TickerCreate(BaseModel):
    """Payload to create a new portfolio ticker."""

    ticker: str = Field(min_length=1, max_length=20)
    company_name: str = Field(min_length=1, max_length=200)
    # Shares must be strictly positive; 4 decimal places stored in DB.
    shares: Decimal = Field(gt=Decimal("0"))
    # Optional cluster assignment at creation time.
    cluster_id: int | None = Field(None, description="ID of the cluster to assign to.")

    @field_validator("ticker", mode="before")
    @classmethod
    def normalise_ticker(cls, v: str) -> str:
        """Force ticker symbols to upper-case on ingestion."""
        return v.strip().upper()

    @field_validator("company_name", mode="before")
    @classmethod
    def strip_company_name(cls, v: str) -> str:
        """Trim surrounding whitespace from company names."""
        stripped = v.strip()
        if not stripped:
            raise ValueError("company_name must not be blank")
        return stripped


class TickerUpdate(BaseModel):
    """Payload to update the share count for an existing ticker."""

    shares: Decimal = Field(gt=Decimal("0"))


class TickerResponse(BaseModel):
    """API response shape for a portfolio ticker."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    ticker: str
    company_name: str
    shares: Decimal

    # Optional cluster assignment.
    cluster_id: int | None = None

    # Market-data fields — None until the first /sync call.
    current_price: Decimal | None = None
    previous_close: Decimal | None = None
    day_change: Decimal | None = None
    day_change_pct: Decimal | None = None
    position_value: Decimal | None = None
    # Rolling 1-year beta vs SPY — None until first sync.
    beta: Decimal | None = None
    synced_at: datetime | None = None

    created_at: datetime
    updated_at: datetime
